### PR TITLE
data filtering poc cont'd

### PR DESCRIPTION
**Pull request by gw000**
Includes grounding fixes

### CODE REVIEW

**Comment by saolsen** [comment]
uh @dhatch, why does it say absolutely every docs file changed?

**Comment by dhatch** [comment]
I think because we added a changelog on main. not certain

File: polar-core/src/debugger.rs [hunk 1]
```diff
@@ -186,7 +184,13 @@ impl Debugger {
             }
         }
         let parts: Vec<&str> = command.split_whitespace().collect();
-        match *parts.get(0).unwrap_or(&"help") {
+        let dflt = match self.last.take() {
```
Comment (dhatch) [inline polar-core/src/debugger.rs:187]: What do these variable names mean? We shouldn't use abbreviations.

File: polar-core/src/debugger.rs [hunk 2]
```diff
@@ -80,6 +83,7 @@ pub struct Debugger {
     /// - `Some(step)`: View the stopping logic in
     ///   [`maybe_break`](struct.Debugger.html#method.maybe_break).
     step: Option<Step>,
+    last: Option<String>,
```
Comment (dhatch) [inline polar-core/src/debugger.rs:86]: What's `last`? Could you add a doc string?

File: polar-core/src/data_filtering.rs [hunk 3]
```diff
@@ -0,0 +1,799 @@
...
+    partial_results: PartialResults,
+    variable: &str,
+    class_tag: &str,
+) -> PolarResult<FilterPlan> {
+    let mut filter_plan = FilterPlan {
+        result_sets: vec![],
+    };
+    // @NOTE(steve): Just reading an env var here sucks (see all the stuff we had to do
+    // to get POLAR_LOG to work in all libs, wasm etc...) but that's what I'm doing today.
+    // At some point surface this info better.
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:719]: +1. It's probably best for us to pass in all these options & read the env vars from the host libraries. If you want you could open a task for this & mark as good for onboarding & cleanup.

Comment (dhatch) [reply polar-core/src/data_filtering.rs:719]: Other aspect of this is having various logging we can turn on and off. See the DEBUG constant at the top of simplifier for another crappy example of this that I added.

File: polar-core/src/data_filtering.rs [hunk 4]
```diff
@@ -0,0 +1,799 @@
...
+        assert_eq!(exp.operator, Operator::And);
+
+        if explain {
+            eprintln!("  {}: {}", i, term.to_polar());
+        }
+
+        let var_info = process_result(exp);
+        let vars = collapse_vars(var_info);
+
+        if explain {
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:739]: Breaking this explain into a separate function would make the main control flow easier to follow.

File: polar-core/src/data_filtering.rs [hunk 5]
```diff
@@ -0,0 +1,799 @@
...
+            }
+        }
+    }
+}
+
+pub type Types = HashMap<String, HashMap<String, Type>>;
+pub type PartialResults = Vec<ResultEvent>;
+
+#[derive(Debug)]
+struct VarInfo {
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:114]: Some comments on these fields & their meaning would be helpful.

File: polar-core/src/data_filtering.rs [hunk 6]
```diff
@@ -0,0 +1,799 @@
...
+    eq_values: Vec<(Symbol, Term)>,                     // x = 1;
+    contained_values: Vec<(Term, Symbol)>,              // 1 in x
+    field_relationships: Vec<(Symbol, String, Symbol)>, // x.a = y
+    in_relationships: Vec<(Symbol, Symbol)>,            // x in y
+}
+
+// @TODO(steve): Better way to handle these checks than just unwraps and asserts.
+
+fn process_result(exp: &Operation) -> VarInfo {
+    let mut var_info = VarInfo {
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:126]: `impl Default`

`let mut var_info = VarInfo::default();`

File: polar-core/src/data_filtering.rs [hunk 7]
```diff
@@ -0,0 +1,799 @@
...
+        Operator::Mod => unimplemented!("`mod` is not supported for data filtering."),
+        Operator::Rem => unimplemented!("`rem` is not supported for data filtering."),
+        Operator::Add => unimplemented!("addition is not supported for data filtering."),
+        Operator::Sub => unimplemented!("subtraction is not supported for data filtering."),
+        Operator::Geq => unimplemented!("`>=` is not supported for data filtering."),
+        Operator::Leq => unimplemented!("`<=` is not supported for data filtering."),
+        Operator::Neq => unimplemented!("`!=` is not supported for data filtering."),
+        Operator::Gt => unimplemented!("`>` is not supported for data filtering."),
+        Operator::Lt => unimplemented!("`<` is not supported for data filtering."),
+        // @TODO(steve): Expand or expressions to multiple bindings in the simplifier.
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:299]: I think a better way to accomplish this would be in the inverter. Then we can drop the Or operator completely from partials.

File: polar-core/src/data_filtering.rs [hunk 8]
```diff
@@ -0,0 +1,799 @@
...
+            } else {
+                unimplemented!("Non pattern specializers are not yet supported for data filtering.")
+            }
+        }
+        Operator::Unify | Operator::Eq | Operator::Assign => {
+            assert_eq!(exp.args.len(), 2);
+
+            let mut lhs = exp.args[0].clone();
+            if let Value::Expression(op) = lhs.value() {
+                lhs = process_exp(var_info, op).unwrap();
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:215]: This means we expect lhs & rhs to be Dot, right?

Could we break `process_exp` into `process_exp_dot`, etc. and only have `process_exp_dot` return a `Term`?

File: polar-core/src/data_filtering.rs [hunk 9]
```diff
@@ -0,0 +1,799 @@
...
+        Operator::Dot => {
+            // Dot operations return a var that can be unified with.
+            // We create a new var to represent the result of the operation.
+            assert_eq!(exp.args.len(), 2);
+            let mut var = exp.args[0].clone();
+            if let Ok(inner_exp) = var.value().as_expression() {
+                if inner_exp.operator != Operator::Dot {
+                    unimplemented!("Operations other than dot nested within a dot are not yet supported for data filtering.")
+                }
+                var = process_exp(var_info, inner_exp).unwrap();
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:181]: Same here w/ `process_exp_dot`

File: polar-core/src/data_filtering.rs [hunk 10]
```diff
@@ -0,0 +1,799 @@
...
+        args,
+    }) = var.value().as_expression()
+    {
+        var = Term::new_temporary(Value::Variable(dot_var(
+            var_info,
+            args[0].clone(),
+            &args[1],
+        )))
+    }
+    // TODO(steve): There's a potential name clash here which would be bad. Works for now.
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:151]: If I have `x.foo > 2 and x.foo < 1` that would be a name clash, right?

File: polar-core/src/data_filtering.rs [hunk 11]
```diff
@@ -0,0 +1,799 @@
...
+    // but should probably generate this var better.
+    let sym = var.value().as_symbol().unwrap();
+    let field_str = field.value().as_string().unwrap();
+    let new_var = Symbol::new(&format!("{}_dot_{}", sym.0, field_str));
+
+    // Record the relationship between the vars.
+    var_info
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:158]: It seems like this function just produces the variable name but it actually also updates `var_info`. Any way to keep this in `process_exp`?

File: polar-core/src/data_filtering.rs [hunk 12]
```diff
@@ -0,0 +1,799 @@
...
+                rhs = process_exp(var_info, op).unwrap();
+            };
+
+            match (lhs.value(), rhs.value()) {
+                // l in r
+                (Value::Variable(l), Value::Variable(r)) => {
+                    var_info.in_relationships.push((l.clone(), r.clone()));
+                }
+                // var in [1, 2, 3]
+                (Value::Variable(_var), _val) => {
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:262]: I don't think this every happens because VM would produce choice points & multiple bindings.

File: polar-core/src/data_filtering.rs [hunk 13]
```diff
@@ -0,0 +1,799 @@
...
+        }
+        // Create new cycle if we couldn't
+        let mut new_cycle = HashSet::new();
+        new_cycle.insert(l);
+        new_cycle.insert(r);
+        joined_cycles.push(new_cycle);
+    }
+
+    let mut next_id = 0;
+    let mut get_id = move || {
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:339]: did not know you could do this

File: polar-core/src/data_filtering.rs [hunk 14]
```diff
@@ -0,0 +1,799 @@
...
+        let id = next_id;
+        next_id += 1;
+        format!("{}", id)
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:342]: why is id not an int?

File: polar-core/src/data_filtering.rs [hunk 15]
```diff
@@ -0,0 +1,799 @@
...
+    };
+
+    // Give each cycle an id
+    let mut variables: HashMap<String, HashSet<Symbol>> = HashMap::new();
+    for joined_cycle in joined_cycles {
+        let id = get_id();
+        variables.insert(id, joined_cycle);
+    }
+
+    // Substitute in relationships
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:352]: @dhatch stopped reading here

**Review by dhatch — COMMENTED** [review]
Ok I started reviewing but I ran out of steam for today. Here's my comments so far which are not too substantial & some may be clean ups you don't want to take.

I haven't gotten through the meat of this which is reviewing `data_filtering.rs` beyond `process_exp`.

File: languages/python/oso/polar/polar.py [hunk 16]
```diff
@@ -247,6 +255,45 @@ def get_class(self, name):
...
+        """
+        # Data filtering.
+        resource = Variable("resource")
+        # Get registered class name somehow
+        class_name = self.host.cls_names[cls]
+        constraint = Expression(
+            "And", [Expression("Isa", [resource, Pattern(class_name, {})])]
+        )
+        results = list(
+            self.query_rule(
```
[the remark this answers is not in the document]
Comment (saolsen) [reply languages/python/oso/polar/polar.py:278]: Yeah one day, not for this first version. You run into the same problem that the validation queries had where you need to run the query from the language side so that external checks all work. We can do this later.

File: languages/python/oso/polar/host.py [hunk 17]
```diff
@@ -20,25 +20,56 @@
...
+            return self.types_get_field(obj, field)
 
         self.get_field = get_field or default_get_field
 
+    # @Q: I'm not really sure what I'm returning here.
+    def types_get_field(self, obj, field):
+        obj_type_name = self.cls_names[obj]
+        if obj_type_name in self.types:
+            obj_type_info = self.types[obj_type_name]
+            if field in obj_type_info:
```
[the remark this answers is not in the document]
Comment (saolsen) [reply languages/python/oso/polar/host.py:53]: It checks the registered type info to see what the type of the field lookup is. I'll add a comment.

[the remark this answers is not in the document]
Comment (saolsen) [reply languages/python/oso/polar/host.py:53]: This also will probably go away later. Eventually the type info will be passed into the core and we wont need this anymore it can all be resolved there (not for this release though).

File: languages/python/oso/polar/query.py [hunk 18]
```diff
@@ -94,7 +95,34 @@ def handle_external_call(self, data):
 
         # Lookup the attribute on the instance.
         try:
-            attr = getattr(instance, attribute)
+            # Check if it's a relationship
+            attr = None
```
[the remark this answers is not in the document]
Comment (saolsen) [reply languages/python/oso/polar/query.py:99]: yep and yep

File: polar-core/src/events.rs [hunk 19]
```diff
@@ -96,3 +96,16 @@ pub enum QueryEvent {
         iterable: Term,
     },
 }
+
+// A struct for just Result Events. Used to pass data back into
+// the core for validation and data filtering.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct ResultEvent {
```
[the remark this answers is not in the document]
Comment (saolsen) [reply polar-core/src/events.rs:103]: Agree.

File: polar-core/src/data_filtering.rs (hunk 3 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:719]: yeah, that is a good onboarding task

File: polar-core/src/data_filtering.rs (hunk 7 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:299]: Yeah, even better. Will make a task, that's a followup thing.

File: polar-core/src/data_filtering.rs (hunk 8 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:215]: Maybe, I'm not sure yet if anything else will have to pass out a term. Will put on my list of cleanups for later if not.

File: polar-core/src/data_filtering.rs (hunk 10 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:151]: No that would be fine. Those are both the same var. The name clash this refers to is if you had `x.foo = 1 and x_dot_foo = 2`. It would be a really weird policy lol but those are actually different vars but I'd end up treating them the same.

File: polar-core/src/data_filtering.rs (hunk 12 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:262]: Right

File: polar-core/src/data_filtering.rs (hunk 10 above)
Comment (dhatch) [reply polar-core/src/data_filtering.rs:151]: Oh I got it

File: polar-core/src/data_filtering.rs (hunk 8 above)
Comment (dhatch) [reply polar-core/src/data_filtering.rs:215]: Sg

File: polar-core/src/data_filtering.rs [hunk 20]
```diff
@@ -0,0 +1,799 @@
...
+    process_exp(&mut var_info, exp);
+    var_info
+}
+
+fn dot_var(var_info: &mut VarInfo, var: Term, field: &Term) -> Symbol {
+    let mut var = var;
+    while let Ok(Operation {
+        operator: Operator::Dot,
+        args,
+    }) = var.value().as_expression()
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:143]: Do we need this loop here & also the recursive call in `process_exp`? I didn't think about this too long but it seems like this might run several times for every part of the expression.

File: polar-core/src/data_filtering.rs [hunk 21]
```diff
@@ -0,0 +1,799 @@
...
+    // Give each cycle an id
+    let mut variables: HashMap<String, HashSet<Symbol>> = HashMap::new();
+    for joined_cycle in joined_cycles {
+        let id = get_id();
+        variables.insert(id, joined_cycle);
+    }
+
+    // Substitute in relationships
+    let mut parent_ids = vec![];
+    'parent_relationships: for (parent, _, _) in &var_info.field_relationships {
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:354]: Note to self: `field_relationships` is (var, field, other_var) (var.field = other_var)

File: polar-core/src/data_filtering.rs [hunk 22]
```diff
@@ -0,0 +1,799 @@
...
+        let mut new_set = HashSet::new();
+        new_set.insert(child.clone());
+        variables.insert(new_id.clone(), new_set);
+        child_ids.push(new_id);
+    }
+
+    // If a.b = c and a.b = d, that means c = d.
+    // @Sorry(steve): Wow, what a loop. Maybe just loop over indexes instead.
+    let mut new_unifies: Vec<(String, String)> = vec![];
+    for (i, ((parent_id1, child_id1), (_, field1, _))) in parent_ids
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:387]: Might call `i` `child_id`

File: polar-core/src/data_filtering.rs [hunk 23]
```diff
@@ -0,0 +1,799 @@
...
+        .iter()
+        .zip(child_ids.iter())
+        .zip(var_info.field_relationships.iter())
+        .enumerate()
+    {
+        for (j, ((parent_id2, child_id2), (_, field2, _))) in parent_ids
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:393]: Are these things actually ids? Aren't they the variable names of the parent & the children still?

File: polar-core/src/data_filtering.rs [hunk 24]
```diff
@@ -0,0 +1,799 @@
...
+                new_unifies.push((child_id1.clone(), child_id2.clone()));
+            }
+        }
+    }
+
+    // @TODO(steve): There are absolutely bugs in here.
+    // If we're turning 0 into 1 and then 0 into 2 it'll just blow up
+    // not correctly turn 0 and 1 into 2. Needs some tests.
+    for (x, y) in &new_unifies {
+        if x != y {
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:410]: Why is this logic different than the original logic you used when building `variables`?

File: polar-core/src/data_filtering.rs [hunk 25]
```diff
@@ -0,0 +1,799 @@
...
+        {
+            if i != j && parent_id1 == parent_id2 && field1 == field2 && child_id1 != child_id2 {
+                // Unify children
+                new_unifies.push((child_id1.clone(), child_id2.clone()));
+            }
+        }
+    }
+
+    // @TODO(steve): There are absolutely bugs in here.
+    // If we're turning 0 into 1 and then 0 into 2 it'll just blow up
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:407]: This comment means if we have `0 = 1` & `0 = 2` which would stem from something like `x.bar = y and x.bar = z and x.bar = a` (although not sure how this case would get out of the simplifier)?

File: polar-core/src/data_filtering.rs [hunk 26]
```diff
@@ -0,0 +1,799 @@
...
+                other_field,
+                ..
+            } = typ
+            {
+                constrain_var(result_set, types, vars, child, other_class_tag);
+
+                // If the constrained child var doesn't have any constraints on it, we don't need to
+                // constrain this var. Otherwise we're just saying field foo in all Foos which
+                // would fetch all Foos and not be good.
+                if let Some(child_result) = result_set.requests.remove(child) {
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:595]: Can we just do this later before we insert the request into the result set on L663? Only insert if there are constraints? Then here we just look up the result set & if it exists we add the In & ref constraint.

File: polar-core/src/data_filtering.rs [hunk 27]
```diff
@@ -0,0 +1,799 @@
...
+                });
+                contributed_constraints = true;
+            }
+            assert!(contributed_constraints);
+        }
+    }
+
+    // Constrain any vars that are `in` this var.
+    // Add their constraints to this one.
+    // @NOTE(steve): I think this is right, but I'm not totally sure.
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:652]: So this is for something like:

`y in _this.bar and y.foo = 2`

File: polar-core/src/data_filtering.rs [hunk 28]
```diff
@@ -0,0 +1,799 @@
...
+    // This might assume that the current var is a relationship of kind "children".
+    for (lhs, rhs) in &vars.in_relationships {
+        if rhs == var_id {
+            constrain_var(result_set, types, vars, lhs, var_type);
+            let in_result_set = result_set.requests.remove(lhs).unwrap();
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:657]: Could you just return a `FetchRequest` from `constrain_var`? Then you don't need this logic here

File: polar-core/src/data_filtering.rs [hunk 29]
```diff
@@ -0,0 +1,799 @@
...
+        let new_id = get_id();
+        let mut new_set = HashSet::new();
+        new_set.insert(var.clone());
+        variables.insert(new_id.clone(), new_set);
+        types.insert(new_id, typ);
+    }
+
+    let mut this_id = String::new();
+    for (id, set) in &variables {
+        if set.contains(&Symbol::new("_this")) {
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:527]: Is it a problem if this isn't found?

File: polar-core/src/data_filtering.rs [hunk 30]
```diff
@@ -0,0 +1,799 @@
...
+        Operator::Or => unimplemented!("`or` is not supported for data filtering."),
+        Operator::ForAll => unimplemented!("`forall` is not supported for data filtering."),
+    }
+    None
+}
+
+#[derive(Debug)]
+struct Vars {
+    variables: HashMap<String, HashSet<Symbol>>,
+    field_relationships: HashSet<(String, String, String)>,
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:309]: If we define `Id` as a type, this would be `HashSet<(Id, String, Id)>` which would make it easier to tell what's going on.

File: polar-core/src/data_filtering.rs [hunk 31]
```diff
@@ -0,0 +1,799 @@
...
+                .filter(|r| r.0 == *parent && r.1 != *field && r.2 == *child)
+            {
+                request.constraints.push(Constraint {
+                    kind: ConstraintKind::Eq,
+                    field: field.clone(),
+                    value: ConstraintValue::Field(eqf.1.clone()),
+                });
+                contributed_constraints = true;
+            }
+            assert!(contributed_constraints);
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:646]: So we'd hit this assert if the relationship wasn't defined rn?

File: polar-core/src/data_filtering.rs [hunk 32]
```diff
@@ -0,0 +1,799 @@
...
+) {
+    // @TODO(steve): Probably should check the type against the var types. I think???
+    let type_def = types
+        .iter()
+        .find(|r| r.0 == var_type)
+        .map(|r| r.1.clone())
+        .unwrap_or_else(HashMap::new);
+
+    let mut request = if result_set.requests.contains_key(var_id) {
+        result_set.requests.remove(var_id).unwrap()
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:568]: wouldn't this branch result in pushing the `var_id` to the resolve order twice?

File: polar-core/src/data_filtering.rs [hunk 33]
```diff
@@ -0,0 +1,799 @@
...
+
+    result_set.requests.insert(var_id.to_string(), request);
+    result_set.resolve_order.push(var_id.to_string());
+}
+
+pub fn opt_pass(filter_plan: &mut FilterPlan, explain: bool) -> bool {
+    let mut optimized = false;
+
+    // Remove duplicate result set in a union.
+    let mut drop_plan = None;
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:671]: How does this happen? If two rules are identical?

File: polar-core/src/data_filtering.rs [hunk 34]
```diff
@@ -0,0 +1,799 @@
...
+}
+
+fn constrain_vars(types: &Types, vars: &Vars, this_type: &str) -> ResultSet {
+    let mut result_set = ResultSet {
+        requests: HashMap::new(),
+        resolve_order: vec![],
+        result_id: vars.this_id.clone(),
+    };
+    constrain_var(&mut result_set, types, vars, &vars.this_id, this_type);
+    result_set
```
Comment (dhatch) [inline polar-core/src/data_filtering.rs:550]: How are we actually checking that everything in Vars is converted into a constraint?

For example if I wrote something like `x.bar = x.y.foo`? My undersatnding is that isn't supported right now, but I think it would just silently not be checked. Is that right?

File: polar-core/src/partial/simplify.rs [hunk 35]
```diff
@@ -597,44 +601,55 @@ impl Simplifier {
...
-            term.value(),
-            Value::Dictionary(_) | Value::Call(_) | Value::List(_) | Value::Expression(_)
-        ) {
-            let value = term.mut_value();
-            match value {
-                Value::Dictionary(dict) => {
-                    for (_, v) in dict.fields.iter_mut() {
-                        self.simplify_term(v, simplify_operation.clone());
-                    }
+        if self.seen.contains(term) {
```
Comment (dhatch) [inline polar-core/src/partial/simplify.rs:604]: I don't have the context to review this, what's this doing?

File: languages/python/oso/polar/data_filtering.py [hunk 36]
```diff
@@ -0,0 +1,148 @@
...
+
+    return Constraint(kind=kind, field=field, value=value)
+
+
+def ground_constraints(polar, results, filter_plan, constraints):
+    for constraint in constraints:
+        if isinstance(constraint.value, Ref):
+            ref = constraint.value
+            constraint.value = results[ref.result_id]
+            if ref.field is not None:
```
Comment (dhatch) [inline languages/python/oso/polar/data_filtering.py:108]: Is it valid for `field` to be None?

**Review by dhatch — COMMENTED** [review]
Okay, I've reviewed everything besides simplifier changes & isa changes so far.

File: polar-core/src/partial/isa_constraint_check.rs [hunk 37]
```diff
@@ -65,75 +74,98 @@ impl IsaConstraintCheck {
...
         }
 
-        // a.b.c vs. d
-        if constraint_path
+        let just_vars = constraint_path.len() == 1
+            && proposed_path.len() == 1
+            && matches!(&constraint.args[0].value().as_symbol(), Ok(Symbol(_)))
+            && matches!(&self.proposed.args[0].value().as_symbol(), Ok(Symbol(_)));
+
+        // FIXME(gw): this logic is hard to follow!
```
Comment (dhatch) [inline polar-core/src/partial/isa_constraint_check.rs:85]: Can you explain what this logic is doing?

File: polar-core/src/data_filtering.rs (hunk 32 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:568]: Hmm yeah, gotta fix resolve order in general, It's not right. Will make that a followup task.

File: polar-core/src/data_filtering.rs (hunk 14 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:342]: Oh there were a few reasons. One was so I could re-use a Vec<(String, String)> for variables and ids but I didn't end up actually doing that. The other is that keys in a `json` map have to be strings so I'd have to convert later if I didn't just do it here.

File: polar-core/src/partial/isa_constraint_check.rs (hunk 37 above)
Comment (gw000) [reply polar-core/src/partial/isa_constraint_check.rs:85]: first exclude a bunch of cases where we don't care about this constraint: it's not an `Isa`, it's not checking variables, it's checking two unrelated variables, or the paths don't match. then check for the cases we do care about: the sides are the same (is subclass check), or we already have a constraint on a prefix of an access path (isa with path check).

the only real difference from before is (supposed to be) that now we check if the "root" variable of an existing constraint is an alias for the root of the proposed constraint, instead of if it's identical.

File: polar-core/src/data_filtering.rs (hunk 23 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:393]: They are ids. I'm for sure rewriting these loops.

File: polar-core/src/data_filtering.rs (hunk 24 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:410]: Because we're dealing with ids now instead of variables. We're merging two different variable sets.

File: polar-core/src/partial/simplify.rs (hunk 35 above)
Comment (gw000) [reply polar-core/src/partial/simplify.rs:604]: see [here](https://www.notion.so/osohq/Simplifier-infinite-recursion-ad0479a286854af687b40f7e00db0876). it's kind of a simple-minded fix, I know there are still expressions that the simplifier will choke on, although I'm not sure how to produce them now. I actually think this might not be needed anymore now that we supply aliases to the simplifier, i'll do some testing

File: polar-core/src/data_filtering.rs (hunk 25 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:407]: Yeah.
I haven't actually seen it so maybe it is true that the simplifier takes care of that. Should write a test to try and hit this.

File: polar-core/src/data_filtering.rs (hunk 26 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:595]: Yeah I think that works.

File: polar-core/src/data_filtering.rs (hunk 27 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:652]: Yeah

File: polar-core/src/data_filtering.rs (hunk 28 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:657]: Yeah maybe.

File: polar-core/src/data_filtering.rs (hunk 31 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:646]: Not sure, that's a good test to see what happens.

File: polar-core/src/data_filtering.rs (hunk 33 above)
Comment (saolsen) [reply polar-core/src/data_filtering.rs:671]: We get this from the builtin roles policy. Not sure why, maybe because there's like a parent check and an ancestor check?

File: languages/python/oso/polar/data_filtering.py (hunk 36 above)
Comment (saolsen) [reply languages/python/oso/polar/data_filtering.py:108]: Yeah

File: polar-core/src/partial/isa_constraint_check.rs (hunk 37 above)
Comment (saolsen) [reply polar-core/src/partial/isa_constraint_check.rs:85]: lol gabe

**Review by dhatch — APPROVED** [review]
